### PR TITLE
Fix qasm output for composite gates

### LINKED ISSUE #3945: circuit_with_custom_gates.qasm() does not print the definition of the custom gate
_Originally posted by @1ucian0 in https://github.com/Qiskit/qiskit-terra/pull/3393#issuecomment-592720851_

> > 3. Can we write the loaded QASM back as the same QASM (with the custom definitions)?
> 
> You mean `circuit.qasm()`? Let's leave the exporter out of this PR for now...
> 
> As a reference, this is the exporter behavior:
> ```
> qasm_string = """OPENQASM 2.0;
>                   include "qelib1.inc";
>                   gate my_other_gate(phi,lambda) q
>                     {u3(asin(cos(phi)/2), phi+pi, lambda/2) q;}
>                   gate my_gate(phi) r
>                     {my_other_gate(phi, phi+pi) r;}
>                   qreg qr[1];
>                   my_gate(pi) qr[0];"""
> circuit = QuantumCircuit.from_qasm_str(qasm_string)
> print(circuit.qasm())
> ```
> ```
> OPENQASM 2.0;
> include "qelib1.inc";
> qreg qr[1];
> my_gate(3.141592653589793) qr[0]; 
> ```

### LINKED ISSUE #4943: Cannot load QASM circuits generated with MCX gates
### Information

- **Qiskit Terra version**: master @ 251930a
- **Python version**: 3.5
- **Operating system**: osx

### What is the current behavior?

```
>>> import qiskit as qk
>>> qc.mcx([0,1],2)
>>> qk.QuantumCircuit.from_qasm_str(qc.qasm())
>>> qk.QuantumCircuit.from_qasm_str(qc.qasm()) == qc
True
>>> qc = qk.QuantumCircuit(4)
>>> qc.mcx([0,1,2], 3)
>>> print(qc.qasm())
OPENQASM 2.0;
include "qelib1.inc";
qreg q[4];
mcx_gray q[0],q[1],q[2],q[3];

>>> qk.QuantumCircuit.from_qasm_str(qc.qasm())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/kevin.krsulichibm.com/q/qiskit-terra/qiskit/circuit/quantumcircuit.py", line 1715, in from_qasm_str
    return _circuit_from_qasm(qasm)
  File "/Users/kevin.krsulichibm.com/q/qiskit-terra/qiskit/circuit/quantumcircuit.py", line 2261, in _circuit_from_qasm
    ast = qasm.parse()
  File "/Users/kevin.krsulichibm.com/q/qiskit-terra/qiskit/qasm/qasm.py", line 54, in parse
    return qasm_p.parse(self._data)
  File "/Users/kevin.krsulichibm.com/q/qiskit-terra/qiskit/qasm/qasmparser.py", line 1070, in parse
    self.parser.parse(data, lexer=self.lexer, debug=self.parse_deb)
  File "/Users/kevin.krsulichibm.com/q/qiskit-terra/lib/python3.5/site-packages/ply/yacc.py", line 333, in parse
    return self.parseopt_notrack(input, lexer, debug, tracking, tokenfunc)
  File "/Users/kevin.krsulichibm.com/q/qiskit-terra/lib/python3.5/site-packages/ply/yacc.py", line 1120, in parseopt_notrack
    p.callable(pslice)
  File "/Users/kevin.krsulichibm.com/q/qiskit-terra/qiskit/qasm/qasmparser.py", line 638, in p_unitary_op_2
    self.verify_as_gate(program[1], program[2])
  File "/Users/kevin.krsulichibm.com/q/qiskit-terra/qiskit/qasm/qasmparser.py", line 130, in verify_as_gate
    + "', line", str(obj.line), 'file', obj.file)
qiskit.qasm.exceptions.QasmError: "Cannot find gate definition for 'mcx_gray', line 4 file "
```

### PR DESCRIPTION
**Pull request by HuangJunye**
### Summary

This PR tries to fix a few issues in generating qasm for composite gates.

Fixes #4943  
Fixes #3945

### Details and comments

See discussions and comments in #4943

### CODE REVIEW

**Review by ajavadia — COMMENTED** [review]
Could you add tests for all cases that were failing before but this PR fixes?
Also a reno releasenote that talks about the bugfix.

**Comment by HuangJunye** [comment]
> Could you add tests for all cases that were failing before but this PR fixes?
> Also a reno releasenote that talks about the bugfix.

Sure. I’ll do that. When is the expected date for 0.17 release? Is it March 14? I still need sometime to fix the MCXGate bug.

**Comment by kdk** [comment]
> Sure. I’ll do that. When is the expected date for 0.17 release? Is it March 14? I still need sometime to fix the MCXGate bug.

Targeting March 29. I've added this to the release as a nice to have, and we can always punt to the next release or split out the parts that are ready.

**Comment by HuangJunye** [comment]
@kdk I had made a few changes in a separate branch. It can now handle undefined children gates (without parameters). I need to figure out a way to avoid duplicate gate definition and handle the ordering. 

I will work on it this weekend and hopefully it will be good enough to be merged in this release.

**Comment by HuangJunye** [comment]
This PR is now ready for review. C3XGate, C4XGate, MCXGate, MCXGrayCode, MCXRecursive and MCXVChain are all working properly now. 

- Corrected names of a few gates: c3x, c4x, c3sqrtx, etc.
- Specified number of control nodes in the mcx gate name instead of a generic 'mcx' to avoid duplicate definition for mcx gates with different number of control nodes.
- It's now able to generate gate definition for children gates and the definition is placed in correct order in the qasm string: children gates first, parents gates later.

**Comment by HuangJunye** [comment]
Sorry I think I am made some mistakes with resolving merging conflicts and it automatically trigger review requests to many people.

File: qiskit/circuit/library/standard_gates/x.py [hunk 1]
```diff
@@ -437,7 +437,7 @@ def __init__(self, label=None, ctrl_state=None, *, angle=None):
             angle (float): DEPRECATED. The angle used in the controlled-U1 gates. An angle of π/8
                 yields the sqrt(X) gates, an angle of π/4 the 3-qubit controlled X gate.
         """
-        super().__init__('c3sx', 4, [], num_ctrl_qubits=3, label=label,
+        super().__init__('c3sqrtx', 4, [], num_ctrl_qubits=3, label=label,
```
Comment (Cryoris) [inline qiskit/circuit/library/standard_gates/x.py:440]: Can we keep the name `c3sx`? Since the square-root of X gate is called SX in Qiskit this would be more consistent.

Comment (HuangJunye) [reply qiskit/circuit/library/standard_gates/x.py:440]: I have no problem of using `c3sx`, in fact I prefer this name as well, but in the gate is called `c3sqrtx` in `qelib1.inc`: https://github.com/Qiskit/qiskit-terra/blob/0c8bb3dbf8d688590431ca79a83ba8aede84ed20/qiskit/qasm/libs/qelib1.inc#L242. It's also present here: https://github.com/Qiskit/qiskit-terra/blob/0c8bb3dbf8d688590431ca79a83ba8aede84ed20/qiskit/qasm/pygments/lexer.py#L56

Comment (Cryoris) [reply qiskit/circuit/library/standard_gates/x.py:440]: Oh, good point. Maybe we can change it in `qelib1.inc`, so that we are consistent overall 🙂

**Comment by HuangJunye** [comment]
## Summary
Fixes qasm output for 
- C3XGate
- C3SXGate
- C4XGate
- DXCGate
- MCPhaseGate
- MSGate
- R3XGate
- RYYGate
- RZXGate
- GMS
- GR
- GRX
- GRY
- MCX
- MCXGrayCode
- MCXRecursive
- MCXVChain

Didn't fix
- Permutation
- Diagonal
- MCMT
- Gate parameters

## Details
### Generate gate definition for composite gates
- support gates with children composite gates
- order gate definition based on the need, i.e. children before parents
- handle composite gates with same name
	- same definition => only one definition
	- different definition => rename to make it unique
- add

### remove the need of handling unitary gates qasm output separately. Now they are handle in the ame way as other gates

### Gate name change in circuit library

- rcccx => rc3x
- mcx => cnx
- mcx_gray => cnx_gray
- mcx_recursive => cnx_recursive
- mcx_vchain => cnx_vchain


### rename c3sqrtx to c3sx in qelib1.inc and qasmparser

### modified tests to follow new qasm output definition order

**Comment by HuangJunye** [comment]
@Cryoris Can you please review the PR? I think it's ready.

File: qiskit/circuit/library/standard_gates/x.py [hunk 2]
```diff
@@ -851,9 +851,19 @@ def __new__(cls, num_ctrl_qubits=None, label=None, ctrl_state=None):
             return gate
         return super().__new__(cls)
 
-    def __init__(self, num_ctrl_qubits, label=None, ctrl_state=None, _name='mcx'):
+    def __init__(self, num_ctrl_qubits, label=None, ctrl_state=None, _name=None):
         """Create new MCX gate."""
         num_ancilla_qubits = self.__class__.get_num_ancilla_qubits(num_ctrl_qubits)
+
+        # if MCXGate is not specified via MCXGrayCode, MCXRecursive, MCXVChain
+        if _name is None:
```
Comment (Cryoris) [inline qiskit/circuit/library/standard_gates/x.py:859]: Should we capture the corner case of `num_ctrl_qubits=0`? 🙂 Or is that covered by the `ControlledGate` superclass, i.e. does it raise an error in `super().init()` below if `num_ctrl_qubits` is 0?

File: qiskit/circuit/quantumcircuit.py [hunk 3]
```diff
@@ -23,6 +23,7 @@
 from collections import OrderedDict, defaultdict
 from typing import Union
 import numpy as np
+import qiskit as qk
```
Comment (Cryoris) [inline qiskit/circuit/quantumcircuit.py:26]: Why do we need this import?

File: qiskit/circuit/quantumcircuit.py [hunk 4]
```diff
@@ -206,6 +207,17 @@ def __init__(self, *regs, name=None, global_phase=0, metadata=None):
...
+        self.qelib1_gate_names = ['barrier', 'measure', 'reset',
+                                  'u3', 'u2', 'u1', 'cx', 'id', 'u0',
+                                  'u', 'p', 'x', 'y', 'z', 'h', 's', 'sdg', 't', 'tdg',
+                                  'rx', 'ry', 'rz',
+                                  'sx', 'sxdg', 'cz', 'cy', 'swap', 'ch', 'ccx', 'cswap',
+                                  'crx', 'cry', 'crz', 'cu1', 'cp', 'cu3', 'csx', 'cu',
+                                  'rxx', 'rzz', 'rccx', 'rc3x', 'c3x', 'c3sx', 'c4x']
+        self.existing_composite_circuits = []
+        self.qasm_string = ""
+        self.qasm_string_temp = ""
```
Comment (Cryoris) [inline qiskit/circuit/quantumcircuit.py:219]: I'd prefer to keep the circuit class clean and not add new properties if possible (in particular not public ones):
* I think `self.qasm_string_temp` is used and not `self.qasm_string`, but it seems that even that variable could be kept locally instead of as a circuit property
* `self.existing_composite_circuits` exists only for caching reasons, right? I'm okay with that for now I think, but could you make it private?
* Do the qelib gates need to be a property or can they also be local?

File: qiskit/circuit/quantumcircuit.py [hunk 5]
```diff
@@ -1290,67 +1301,68 @@ def qasm(self, formatted=False, filename=None):
...
+            code = pygments.highlight(self.qasm_string_temp,
                                       OpenQASMLexer(),
                                       Terminal256Formatter(style=QasmTerminalStyle))
             print(code)
             return None
         else:
-            return string_temp
+            return self.qasm_string_temp
+
+    def _insert_composite_gate_definition_qasm(self):
```
Comment (Cryoris) [inline qiskit/circuit/quantumcircuit.py:1344]: Maybe this, as well as `_get_composite_circuit_qasm_from_instruction` can be private module-level functions (same file as `QuantumCircuit` but not inside the class) to keep the code cleaner and easier to read. That would mean that instead of saving the qasm string to an internal variable you would have to pass it around. That's arguably a bit more effort to program but I think it's easier to maintain and read 🙂

**Review by Cryoris — COMMENTED** [review]
Thanks for the effort Junye! The direction looks good, I think it just needs a bit more cleanup, in particular to only add circuit properties if they are needed. I left some comments below 🙂

**Comment by ajavadia** [comment]
@HuangJunye any update here? If you can address review comments from @Cryoris and update the branch then I hope we can include this in the release for June 22nd.

**Comment by HuangJunye** [comment]
@ajavadia I am a bit tied up this week. Is it too late to meet the next release if I address the comments by next Tuesday (June 15)?

**Comment by ajavadia** [comment]
@HuangJunye I think that would be fine.

File: qiskit/circuit/quantumcircuit.py (hunk 3 above)
Comment (HuangJunye) [reply qiskit/circuit/quantumcircuit.py:26]: I can't remember why I added it :D It's now removed.

File: qiskit/circuit/quantumcircuit.py (hunk 4 above)
Comment (HuangJunye) [reply qiskit/circuit/quantumcircuit.py]: Let me check. I think they can be local variables. Sorry I am not an expert in python or programming in general. These might be silly questions:
- To make them local variables, should I just remove `self.` and use the variables name directly? 
- To make `self.existing_composite_circuits` "private, should I just add `_`, like `self._existing_composite_circuits`?
- I will remove `self.qasm_string` upon checking that it is not used.

File: qiskit/circuit/quantumcircuit.py (hunk 5 above)
Comment (HuangJunye) [reply qiskit/circuit/quantumcircuit.py:1785]: Sure, let me work on that.

File: qiskit/circuit/quantumcircuit.py [hunk 6]
```diff
@@ -45,9 +47,17 @@
...
-    from pygments.formatters import Terminal256Formatter  # pylint: disable=no-name-in-module
+    from pygments.formatters import (
+        Terminal256Formatter,
+    )  # pylint: disable=no-name-in-module
     from qiskit.qasm.pygments import OpenQASMLexer  # pylint: disable=ungrouped-imports
+<<<<<<< HEAD
+    from qiskit.qasm.pygments import (
+        QasmTerminalStyle,
+    )  # pylint: disable=ungrouped-imports
+=======
```
Comment (HuangJunye) [inline qiskit/circuit/quantumcircuit.py:58]: @1ucian0 There's a lot of merge conflicts in this file. Should I resolve them?

Comment (1ucian0) [reply qiskit/circuit/quantumcircuit.py:58]: Ups. I missed then. Please go ahead

File: qiskit/circuit/quantumcircuit.py (hunk 4 above)
Comment (Cryoris) [reply qiskit/circuit/quantumcircuit.py]: Yes to all of the above 😄

File: qiskit/circuit/library/standard_gates/x.py (hunk 2 above)
Comment (1ucian0) [reply qiskit/circuit/library/standard_gates/x.py:888]: I think it's in `qiskit/circuit/controlledgate.py`:
```
CircuitError: 'The number of control qubits must be in [1, num_qubits-1]'
```

**Review by 1ucian0 — DISMISSED** [review]
Do you mind adding a release note?

File: qiskit/circuit/quantumcircuit.py [hunk 7]
```diff
@@ -1445,60 +1417,32 @@ def qasm(self, formatted=False, filename=None, encoding=None):
             if instruction.name == "measure":
                 qubit = qargs[0]
                 clbit = cargs[0]
-                string_temp += "{} {} -> {};\n".format(
+                string_temp += "{} {} -> {};\n".forma(
```
Comment (1ucian0) [inline qiskit/circuit/quantumcircuit.py:1420]: This typo means that this code is not being covered in the tests. 
```suggestion
                string_temp += "{} {} -> {};\n".format(
```
Maybe add a test for it?

File: qiskit/circuit/quantumcircuit.py [hunk 8]
```diff
@@ -1352,55 +1325,56 @@ def qasm(self, formatted=False, filename=None, encoding=None):
                 ``True``.
             QasmError: If circuit has free parameters.
         """
-        from qiskit.circuit.controlledgate import ControlledGate
 
         if self.num_parameters > 0:
             raise QasmError("Cannot represent circuits with unbound parameters in OpenQASM 2.")
 
         existing_gate_names = [
```
Comment (1ucian0) [inline qiskit/circuit/quantumcircuit.py:1332]: This list should only include the names in of gates in `qelib1.inc` (p10-12 of https://arxiv.org/abs/1707.03429)

File: qiskit/circuit/quantumcircuit.py [hunk 9]
```diff
@@ -1445,60 +1417,32 @@ def qasm(self, formatted=False, filename=None, encoding=None):
...
-                )
             else:
-                string_temp += "{} {};\n".format(
+                # decompose gate using definitions if they are not defined in OpenQASM2
+                if instruction.name not in existing_gate_names:
+                    if instruction not in existing_composite_circuits:
+                        custom_instr_names = [
+                            instruction.name for instruction in existing_composite_circuits
+                        ]
+                        if instruction.name in custom_instr_names:
```
Comment (1ucian0) [inline qiskit/circuit/quantumcircuit.py:1432]: Nested conditionals can be a combined with `and`:
```suggestion
                if instruction.name not in existing_gate_names and instruction not in existing_composite_circuits and instruction.name in [instruction.name for instruction in existing_composite_circuits]:
```

File: qiskit/circuit/quantumcircuit.py [hunk 10]
```diff
@@ -23,6 +23,7 @@
 from collections import OrderedDict, defaultdict
 from typing import Union
 import numpy as np
+import qiskit
```
Comment (1ucian0) [inline qiskit/circuit/quantumcircuit.py:26]: is this needed?

File: qiskit/circuit/quantumcircuit.py (hunk 7 above)
Comment (HuangJunye) [reply qiskit/circuit/quantumcircuit.py]: Thanks for catching and fixing this. I have not idea why my VSCode didn't complain.

File: qiskit/circuit/quantumcircuit.py (hunk 8 above)
Comment (HuangJunye) [reply qiskit/circuit/quantumcircuit.py:1321]: Yes, but more gates definitions have been added to `qelib1.inc` inside qiskit code itself so all those gates can be parsed: https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/qasm/libs/qelib1.inc

Comment (HuangJunye) [reply qiskit/circuit/quantumcircuit.py:1321]: I just saw your new PR about reverting to the paper definition: https://github.com/Qiskit/qiskit-terra/pull/6125 I will change my code.

Comment (HuangJunye) [reply qiskit/circuit/quantumcircuit.py:1321]: `existing_gate_names` need to match with those defined here: https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/qasm/libs/qelib1.inc otherwise it may produce `Duplicate declaration` error. So my PR is dependent on yours: #6125

Comment (HuangJunye) [reply qiskit/circuit/quantumcircuit.py:1321]: Can I keep the extra gates definition for now and make another PR when this one #6125 gets merged?

File: qiskit/circuit/quantumcircuit.py (hunk 10 above)
Comment (HuangJunye) [reply qiskit/circuit/quantumcircuit.py:26]: No needed any more. Let me remove it.

**Comment by HuangJunye** [comment]
@1ucian0 Thanks a lot for helping me fix the PR today. All failed tests now passed locally. Waiting for CI testing. I also added the release note. Can you please take a look? Hope it can make it for 0.18.

cc: @mtreinish

**Comment by HuangJunye** [comment]
Finally, all tests are passed! Thanks a ton @1ucian0 !!!

File: qiskit/circuit/quantumcircuit.py [hunk 11]
```diff
@@ -3044,3 +2979,81 @@ def _compare_parameters(param1, param2):
...
+
+    # Cycle through all gate definitions and add all undefined gates to the list
+    for instruction in existing_composite_circuits:
+        _get_composite_circuit_qasm_from_instruction(
+            instruction, existing_gate_names, existing_composite_circuits
+        )
+
+    # Generate gate definition string
+    for instruction in existing_composite_circuits:
+        if hasattr(instruction, "_qasm_definition"):
```
Comment (Cryoris) [inline qiskit/circuit/quantumcircuit.py:3048]: Nothing to change, just for me to understand: only the `UnitaryGate` has the `_qasm_definition` attribute, right?

File: qiskit/circuit/quantumcircuit.py [hunk 12]
```diff
@@ -3044,3 +2979,81 @@ def _compare_parameters(param1, param2):
...
+):
+    """Insert composite gate definition QASM code right after extension library in the header"""
+
+    gate_definition_string = ""
+
+    # Cycle through all gate definitions and add all undefined gates to the list
+    for instruction in existing_composite_circuits:
+        _get_composite_circuit_qasm_from_instruction(
+            instruction, existing_gate_names, existing_composite_circuits
+        )
```
Comment (Cryoris) [inline qiskit/circuit/quantumcircuit.py:3044]: I'm a bit confused about this block, why is this needed? It seemed to me that the next block (L3047-3054) iterates over the same set of instructions 🤔

File: qiskit/circuit/quantumcircuit.py [hunk 13]
```diff
@@ -1451,55 +1409,30 @@ def qasm(self, formatted=False, filename=None, encoding=None):
...
+                # decompose gate using definitions if they are not defined in OpenQASM2
+                if (
+                    instruction.name not in existing_gate_names
+                    and instruction not in existing_composite_circuits
+                ):
+                    if instruction.name in [
+                        instruction.name for instruction in existing_composite_circuits
+                    ]:
+                        # append instruction id to name to make it unique
+                        instruction.name += f"_{id(instruction)}"
```
Comment (Cryoris) [inline qiskit/circuit/quantumcircuit.py:1422]: This seems to modify the circuit data, is that correct? If yes, can we avoid that?

**Review by Cryoris — COMMENTED** [review]
So in a nutshell, the approach is to keep a big list of all unique instructions and then adding definitions for those instructions that are not in the standard set, right?

I left some few comments below and I have one question: is this adding of definitions expensive? Do we want to cache something? Probably not, but I just wanted to be sure we thought about that 🙂

File: qiskit/circuit/quantumcircuit.py (hunk 11 above)
Comment (HuangJunye) [reply qiskit/circuit/quantumcircuit.py:3048]: I think so. This part was modified by @1ucian0. My original code was checking the instance of `UnitaryGate`: https://github.com/Qiskit/qiskit-terra/pull/5887/commits/a9e80dcab2b6af4c920aa6ae1ecb5d7f0c54b9b8

Comment (1ucian0) [reply qiskit/circuit/quantumcircuit.py:3048]: Correct. Maybe we should have a standard way for a `Gate` to express "here is my QASM definition".

File: qiskit/circuit/quantumcircuit.py (hunk 13 above)
Comment (HuangJunye) [reply qiskit/circuit/quantumcircuit.py:1422]: Indeed circuit data is modified here. I think the problem is we are using name to identify gates / instructions. We can't distinguish gates if they have the same name but different definitions. I can revert this and leave the handling of gates with same name out of this PR. What do you think?

File: qiskit/circuit/quantumcircuit.py (hunk 12 above)
Comment (HuangJunye) [reply qiskit/circuit/quantumcircuit.py:3044]: There's probably better way to structure the code. Basically this loop is used to add all gates that are not in `existing_composite_circuits` to `existing_composite_circuits` list, while the loop below is to add qasm definition string to the output qasm string.

File: qiskit/circuit/quantumcircuit.py (hunk 13 above)
Comment (Cryoris) [reply qiskit/circuit/quantumcircuit.py:1422]: We need the instructions in `existing_composite_circuits` to have unique names right? So we could just store a copy of the instruction where we modified the name instead of storing the same instance that lives in circuit.data 🙂

File: qiskit/circuit/quantumcircuit.py (hunk 12 above)
Comment (Cryoris) [reply qiskit/circuit/quantumcircuit.py:3044]: Would it be simpler to do either this in one go or to have two separate functions for this?

Comment (HuangJunye) [reply qiskit/circuit/quantumcircuit.py:3044]: Yeah, now I think about it, I think it should be two separate functions. Let see if I can come up with a quick solution.

File: qiskit/circuit/quantumcircuit.py (hunk 13 above)
Comment (HuangJunye) [reply qiskit/circuit/quantumcircuit.py:1422]: Yes, you are right. Let me fix this.

Comment (HuangJunye) [reply qiskit/circuit/quantumcircuit.py:1422]: I see that this is the same as the original code. Can I leave this part out of this PR?

Comment (Cryoris) [reply qiskit/circuit/quantumcircuit.py:1422]: Yes 👍🏻 could you open an issue for it though?

Comment (HuangJunye) [reply qiskit/circuit/quantumcircuit.py:1422]: Sure, I'll do that.

Comment (HuangJunye) [reply qiskit/circuit/quantumcircuit.py:1422]: Issue created here: https://github.com/Qiskit/qiskit-terra/issues/6687 Please add comments if you think the description is not accurate.

File: qiskit/circuit/quantumcircuit.py (hunk 12 above)
Comment (HuangJunye) [reply qiskit/circuit/quantumcircuit.py]: I separated this function into two: `_add_sub_instruction_to_existing_composite_circuits` and `_get_composite_circuit_qasm_from_instruction` in the latest two commits. Can you please have a look? I hope the code now looks clearer and easier to understand.

File: qiskit/circuit/quantumcircuit.py (hunk 13 above)
Comment (Cryoris) [reply qiskit/circuit/quantumcircuit.py:1422]: Perfect, thanks a lot!

File: test/python/circuit/test_circuit_qasm.py [hunk 14]
```diff
@@ -174,6 +174,35 @@ def test_circuit_qasm_with_multiple_composite_circuits_with_same_name(self):
...
 
+    def test_circuit_qasm_with_composite_circuit_with_children_composite_circuit(self):
+        """Test circuit qasm() method when composite circuits with children
+        composite circuits in the definitions are added to the circuit"""
+
+        child_circ = QuantumCircuit(2, name="child_circ")
+        child_circ.h(0)
+        child_circ.cx(0, 1)
+
+        parent_circ = QuantumCircuit(3, name="parant_circ")
```
Comment (Cryoris) [inline test/python/circuit/test_circuit_qasm.py:185]:
```suggestion
        parent_circ = QuantumCircuit(3, name="parent_circ")
```

File: test/python/circuit/test_circuit_qasm.py [hunk 15]
```diff
@@ -174,6 +174,35 @@ def test_circuit_qasm_with_multiple_composite_circuits_with_same_name(self):
...
+        grandparent_circ.x(3)
+
+        qc = QuantumCircuit(4)
+        qc.append(grandparent_circ, range(4))
+
+        expected_qasm = """OPENQASM 2.0;
+include "qelib1.inc";
+gate child_circ q0,q1 { h q0; cx q0,q1; }
+gate parant_circ q0,q1,q2 { child_circ q0,q1; h q2; }
+gate grandparent_circ q0,q1,q2,q3 { parant_circ q0,q1,q2; x q3; }
```
Comment (Cryoris) [inline test/python/circuit/test_circuit_qasm.py:200]:
```suggestion
gate grandparent_circ q0,q1,q2,q3 { parent_circ q0,q1,q2; x q3; }
```

**Review by Cryoris — APPROVED** [review]
LGTM, thanks for the improvement Junye!

One thing we could change in the future is how the loaded circuits are draw, e.g. dumping and loading the following circuit
```
        ┌───┐┌───┐         ░ ┌─┐             ░
   q_0: ┤ H ├┤ X ├─────────░─┤M├─────────────░───■──
        ├───┤└─┬─┘         ░ └╥┘┌─┐          ░   │
   q_1: ┤ H ├──■───────────░──╫─┤M├──────────░───┼──
        ├───┤  │           ░  ║ └╥┘┌─┐       ░   │
   q_2: ┤ H ├──■───────────░──╫──╫─┤M├───────░───o──
        ├───┤  │  ┌──────┐ ░  ║  ║ └╥┘┌─┐    ░   │
   q_3: ┤ H ├──■──┤0     ├─░──╫──╫──╫─┤M├────░───┼──
        └───┘     │  sub │ ░  ║  ║  ║ └╥┘┌─┐ ░ ┌─┴─┐
   q_4: ──────────┤1     ├─░──╫──╫──╫──╫─┤M├─░─┤ X ├
                  └──────┘ ░  ║  ║  ║  ║ └╥┘ ░ └───┘
meas: 5/══════════════════════╩══╩══╩══╩══╩═════════
```
yields 
```
        ┌───┐┌──────┐         ░ ┌─┐             ░ ┌─────────┐
   q_0: ┤ H ├┤3     ├─────────░─┤M├─────────────░─┤0        ├
        ├───┤│      │         ░ └╥┘┌─┐          ░ │         │
   q_1: ┤ H ├┤0     ├─────────░──╫─┤M├──────────░─┤         ├
        ├───┤│  mcx │         ░  ║ └╥┘┌─┐       ░ │         │
   q_2: ┤ H ├┤1     ├─────────░──╫──╫─┤M├───────░─┤1 ccx_o1 ├
        ├───┤│      │┌──────┐ ░  ║  ║ └╥┘┌─┐    ░ │         │
   q_3: ┤ H ├┤2     ├┤0     ├─░──╫──╫──╫─┤M├────░─┤         ├
        └───┘└──────┘│  sub │ ░  ║  ║  ║ └╥┘┌─┐ ░ │         │
   q_4: ─────────────┤1     ├─░──╫──╫──╫──╫─┤M├─░─┤2        ├
                     └──────┘ ░  ║  ║  ║  ║ └╥┘ ░ └─────────┘
meas: 5/═════════════════════════╩══╩══╩══╩══╩═══════════════
```

**Review by Cryoris — APPROVED** [review]

**Comment by 1ucian0** [comment]
> One thing we could change in the future is how the loaded circuits are draw, e.g. dumping and loading the following circuit

I think this is expected. Qiskit has no warranties about the fidelity of QASM as a serialization format.

**Review by 1ucian0 — APPROVED** [review]

**Comment by HuangJunye** [comment]
@Cryoris @1ucian0 Thanks for the approval!!! I am so happy that this is finally going to be merged!
